crates/services/producer/src/block_producer.rs

```diff
@@ -203,25 +202,28 @@ where
         let header = self._new_header(height, Tai64::now())?;
         let component = Components {
             header_to_produce: header,
-            transactions_source: transaction,
+            transactions_source: transactions.clone(),
             gas_limit: u64::MAX,
         };
 
         let executor = self.executor.clone();
+
         // use the blocking threadpool for dry_run to avoid clogging up the main async runtime
-        let res: Vec<_> =
-            tokio_rayon::spawn_fifo(move || -> anyhow::Result<Vec<Receipt>> {
-                Ok(executor
-                    .dry_run(component, utxo_validation)?
-                    .into_iter()
-                    .flatten()
-                    .collect())
-            })
-            .await?;
-        if is_script && res.is_empty() {
-            return Err(anyhow!("Expected at least one set of receipts"))
+        let tx_statuses = tokio_rayon::spawn_fifo(
+            move || -> anyhow::Result<Vec<TransactionExecutionStatus>> {
+                Ok(executor.dry_run(component, utxo_validation)?)
+            },
+        )
+        .await?;
+
+        for (transaction, tx_status) in transactions.iter().zip(tx_statuses.iter()) {
```

Comment:
Nit: If you want to save yourself the `clone` of `transations` above, you could just `iter` over the `transactions` earlier to get the `is_script` for each of them. I'm only guessing that the `clone` is more expensive than the second loop of `transactions`.

Reply:
Very Nit: Then this for-loop could be replaced with:
```
let with_empty_reciepts = tx_statuses.iter().map(|status| status.receipts.is_empty());
if transactions.iter().zip(with_empty_receipts).any(|(is_script, is_empty)| is_script && is_empty) {
    return Err(anyhow!("Expected at least one set of receipts"))
}
```